src/Common/HashTable/TimeBucketHashTable.h

```diff
@@ -438,4 +454,45 @@ class TimeBucketHashTable : private boost::noncopyable, protected Hash /// empty
 
         return buckets;
     }
+
+    bool isUpdatedBucket(Int64 bucket_) const
+    {
+        auto it = bucket_updated_flags.find(bucket_);
+        if (it != bucket_updated_flags.end())
+            return it->second;
+
+        return false;
+    }
+
+    void resetUpdated(Int64 bucket_)
+    {
+        auto it = bucket_updated_flags.find(bucket_);
+        if (it != bucket_updated_flags.end())
+            it->second = false;
+    }
+
+    void writeBucketUpdatedFlags(DB::WriteBuffer & wb) const
+    {
+        DB::writeVarUInt(bucket_updated_flags.size(), wb);
+        for (const auto & [bucket, updated] : bucket_updated_flags)
+        {
+            DB::writeIntBinary(bucket, wb);
```

Comment:
Let's use `DB::writeVarInt / VarUInt / writeBinary` where possible and avoid full binary / text serde since they are slower and bigger